indexing/src/main/java/com/google/enterprise/cloudsearch/sdk/indexing/IncludeExcludeFilter.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.enterprise.cloudsearch.sdk.indexing;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableList;
+import com.google.enterprise.cloudsearch.sdk.InvalidConfigurationException;
+import com.google.enterprise.cloudsearch.sdk.config.Configuration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Helper utility to check if a particular item should be indexed. Include/exclude
+ * patterns are specified using Java regular expressions; see <a
+ * href="https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html">java.util.regex.Pattern</a>.
+ *
+ * <p>A connector can choose whether to use this helper; see the connector documentation
+ * to find out if this is supported.
+ *
+ * <p>This filter uses configuration property names with the following patterns:
+ * <ul>
+ *   <li>include patterns are prefixed with "includeExcludeFilter.include.regex."
+ *   <li>exclude patterns are prefixed with "includeExcludeFilter.exclude.regex."
+ * </ul>
```

Comment:
I'm not sure if this is the right place, but we should add something like "Allowed values much match at least one include pattern, if any are given, and not match any exclude patterns."